### DISCUSSION TITLE
[Feature] Support Prometheus metrics

### DISCUSSION THREAD

**Discussion by mrueg** [rspamd/rspamd#3484] [Ideas]
## Summary

Currently rspamd only supports graphite metrics, it would be great if there were support for https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format metrics as well
## Motivation
We use prometheus to monitor systems and would like to monitor rspamd as well.

## Describe alternatives you've considered

Setting up graphite and pull via prometheus from there is not viable as it requires additional services to maintain.

## Additional context

Some implementations how this could look like as a lua-plugin:
https://github.com/knyar/nginx-lua-prometheus/blob/master/prometheus.lua
https://github.com/tarantool/prometheus

**Comment by vstakhov** [comment]
Tarantool library looks almost ready to use. However, I have neither need nor desire to write such a plugin. But I leave space for volounteers here. 

The current metric_exporter plugin is poorly designed and needs to be rewritten. I even had some branch with this work but it has never been finished.

**Comment by fatalbanana** [comment]
One might query `/stat` on the controller directly I suppose- but the result is not suitable for direct consumption by Prometheus. One could write an `exporter` to deal with it and scrape that instead; it looks like others are using [json-exporter](https://github.com/shiroyagicorp/prometheus-json-exporter) with rspamd so far.

Native support could we exposed via something like `/stat?format=prometheus` if implemented.

**Comment by dragoangel** [comment]
Yep, native support will be cool. For example HAproxy had csv export that was been converted to prometheus by dedicated exporter, but after some time HAproxy created buildin exported, and it quite popular as far I know

**Comment by mrueg** [comment]
For those interested in getting prometheus metrics now, @WilliButz has implemented a module using json-exporter for NixOS (https://github.com/NixOS/nixpkgs/pull/68496) and shared a rendered config here: https://gist.github.com/WilliButz/2ed6cb59eb7e684ed79a5b4b09503ccc

**Comment by systemcrash** [comment]
Frustrating. Didn't realize /stat was available. Doesn't seem like a stretch to go from JSON to Prometheus metrics. 


> The current metric_exporter plugin is poorly designed and needs to be rewritten. I even had some branch with this work but it has never been finished.

Which branch, BTW?

Let us put it this way: once a prometheus exporter is implemented, this repo and project now have keywords [Prometheus and exporter](https://trends.google.com/trends/explore?date=today%205-y&q=prometheus%20monitoring). 

People like to monitor their key infrastructure. 😄

**Comment by systemcrash** [comment]
Also, here's [roughly how openwrt do it](https://github.com/openwrt/packages/tree/openwrt-19.07/utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors).

**Comment by dragoangel** [comment]
Here is link to rspamd-exporter https://hub.docker.com/r/badsmoke/rspamd-exporter which convert current stats in prometheus format

**Comment by mrueg** [comment]
This is now supported:
https://github.com/rspamd/rspamd/pull/3754
https://github.com/rspamd/rspamd/commit/1cf0d5687baf976760681b0db6eed754555ca89b

**Comment by Jean-Daniel** [comment]
This endpoint is welcome, but it should follow [OpenMetrics naming guidelines](https://prometheus.io/docs/practices/naming/).

**Comment by ivanov17** [comment]
I'm curious why the decision was made to implement the `/metrics` endpoint on the controller worker.

I think it would make more sense to use the normal worker for this. The controller worker may not be used at all since WebUI may not be needed and the normal worker can be used to learn messages. But in this case, people will especially need metrics. 

It also seems a little strange to not have a choice whether to password protect metrics or not, while at the same time having the same password for metrics and other read-only endpoints.

**Reply by dragoangel** [reply]
Generally metrics usually should be pki protected, like required to provide user cert. Password is also an option, but not common for metrics. I understand why the controller was used here, because of the way the data controller operates. You generally can't run normal without a controller process as many things would break anyway. You can run proxy or fuzzy without a controller through...

I personally prefer to write data to elastic, it's provide much more details plus it's possible to create prom metrics from elastic queries as well.

**Reply by ivanov17** [reply]
Thank you for the reply. My mail server setup is currently being tested, but I don't see any problem despite the rspamd controller worker being disabled. I would rather get the metrics and use Grafana for visualization instead of using the native WebUI. 

I'm also concerned about the `secure_ip` setting, which allows all commands from selected addresses. I wouldn't want this behavior even on localhost. Unfortunately, I don't see an option that would, for example, set a list of endpoints that will always be password protected, or vice versa. I expect more problems with the controller worker enabled than with it disabled.

Unfortunately, elastic is not an option for me. I'm going to try get rspamd metrics using graphite port and prometheus graphite exporter. But it looks like rspamd unfortunately provides fewer metrics for graphite. Does the graphite plugin get metrics in a different way than the prometheus endpoint? Or maybe the documentation is a little out of date?

**Comment by vstakhov** [comment]
I see no problems in provinding `/metrics` endpoint to the normal worker as this all statisitics is shared.

**Comment by vstakhov** [comment]
The only problematic API calls are fuzzy and bayes stats. For now, I will limit them to the controller only, however, their architecture should be improved for sure.
